src/compiler/moduleNameResolver.ts

```diff
@@ -816,6 +818,33 @@ namespace ts {
         return loadNodeModuleFromDirectory(extensions, candidate, failedLookupLocations, onlyRecordFailures, state, considerPackageJson);
     }
 
+    function parseNodeModuleFromPath(path: string): { packageDirectory: string, subModuleName: string } | undefined {
+        const nodeModules = "/node_modules/";
+        const idx = path.indexOf(nodeModules);
+        if (idx === -1) {
+            return undefined;
+        }
+
+        const indexAfterNodeModules = idx + nodeModules.length;
+        let indexAfterPackageName = indexOfNextSlash(path, indexAfterNodeModules);
+        if (path.charCodeAt(indexAfterNodeModules) === CharacterCodes.at) {
+            indexAfterPackageName = indexOfNextSlash(path, indexAfterPackageName);
+        }
+        const packageDirectory = path.slice(0, indexAfterPackageName);
+        const subModuleName = removeExtensionAndIndex(path.slice(indexAfterPackageName));
+        return { packageDirectory, subModuleName };
+    }
+
+    function indexOfNextSlash(s: string, pos: number): number {
+        const x = s.indexOf(directorySeparator, pos);
+        return x === -1 ? pos : x;
+    }
+
+    function removeExtensionAndIndex(fileName: string): string {
+        const noExtension = removeFileExtension(fileName);
+        return noExtension === "index" ? "" : removeSuffix(noExtension, "/index");

```

Comment:
Presumably, we won't reach here for (e.g.) "index.vb"?

Reply:
`removeFileExtension` only removes the supported file extensions (and also correctly removes all of `.d.ts` and not just `.ts`. Although it's also true that `fileName` shouldn't have a `.vb` extension since we'd never resolve a module to a `.vb` file.